File: codes/matlab/ch10/poisson2d_cheb.m
%% poisson2d_cheb.m
%
% Figure 10.7: 2D Poisson equation on Chebyshev tensor grid
%
% PDE: -Lap(u) = f, -1 < x, y < 1
% BCs: u = 0 on boundary (Dirichlet)
%
% Uses manufactured solution:
%   u(x, y) = (1 - x^2)(1 - y^2) cos(pi*x/2) cos(pi*y/2)
%
% Author: Dr. Denys Dutykh (Khalifa University, Abu Dhabi, UAE)
% Part of "Computational Etudes: A Spectral Approach"
% https://github.com/dutykh/computational-etudes

clear; close all; clc;

%% Configuration
N = 32;

% Colors
NAVY = [0.078 0.176 0.431];

% Output path
script_dir = fileparts(mfilename('fullpath'));
output_dir = fullfile(script_dir, '..', '..', '..', 'textbook', 'figures', 'ch10', 'matlab');
output_file = fullfile(output_dir, 'poisson2d_solution.pdf');

if ~exist(output_dir, 'dir')
    mkdir(output_dir);
end

%% Exact solution (manufactured)
exact_solution = @(x, y) (1 - x.^2) .* (1 - y.^2) .* cos(pi*x/2) .* cos(pi*y/2);

%% Setup grid and operators
[D, x] = cheb_matrix(N);
D2 = D * D;
[xx, yy] = meshgrid(x, x);

% Exact solution
U_exact = exact_solution(xx, yy);

%% Interior block
D2_int = D2(2:N, 2:N);
I_int = eye(N-1);

% Kronecker sum Laplacian (represents Δ)
L = kron(D2_int, I_int) + kron(I_int, D2_int);

% For -Δu = f, we have -L * u = f, so L * u = -f
% Or equivalently, use A = -L so that A * u = f
A = -L;

%% RHS from analytical Laplacian (not discrete, to test convergence)
f_full = -exact_laplacian(xx, yy);
f_int = f_full(2:N, 2:N);

%% Solve
u_vec = A \ f_int(:);

%% Reconstruct full solution
U = zeros(N+1, N+1);
U(2:N, 2:N) = reshape(u_vec, N-1, N-1);

%% Compute error
err = abs(U - U_exact);
max_error = max(err(:));
fprintf('N = %d, Max error = %.2e\n', N, max_error);

%% Convergence study upfront so panel (d) can plot it
fprintf('\nConvergence Study:\n');
fprintf('----------------------------------------\n');
fprintf('%6s  %14s  %10s\n', 'N', 'Max Error', 'Ratio');
fprintf('----------------------------------------\n');

N_values = [8, 12, 16, 20, 24, 28, 32, 40, 48];
errs_conv = zeros(size(N_values));
prev_error = NaN;

for i = 1:length(N_values)
    Ni = N_values(i);
    [Di, xi] = cheb_matrix(Ni);
    D2i = Di * Di;
    [xxi, yyi] = meshgrid(xi, xi);

    U_exact_i = exact_solution(xxi, yyi);
    D2_int_i = D2i(2:Ni, 2:Ni);
    I_int_i = eye(Ni-1);
    Li = kron(D2_int_i, I_int_i) + kron(I_int_i, D2_int_i);
    Ai = -Li;

    f_full_i = -exact_laplacian(xxi, yyi);
    f_int_i = f_full_i(2:Ni, 2:Ni);

    u_vec_i = Ai \ f_int_i(:);
    Ui = zeros(Ni+1, Ni+1);
    Ui(2:Ni, 2:Ni) = reshape(u_vec_i, Ni-1, Ni-1);

    erri = max(abs(Ui(:) - U_exact_i(:)));
    errs_conv(i) = erri;

    if ~isnan(prev_error) && erri > 1e-15
        ratio = prev_error / erri;
    else
        ratio = NaN;
    end

    fprintf('%6d  %14.6e  %10.2f\n', Ni, erri, ratio);
    prev_error = erri;
end
fprintf('----------------------------------------\n');

%% Create figure with 4 panels (2x2)
fig = figure('Units', 'inches', 'Position', [1, 1, 11, 9]);

vmin = min(U_exact(:));
vmax = max(U_exact(:));

% Panel (a): Exact solution
subplot(2, 2, 1);
contourf(xx, yy, U_exact, 30, 'LineStyle', 'none');
hold on;
contour(xx, yy, U_exact, 10, 'k', 'LineWidth', 0.3);
xlabel('x', 'FontSize', 11);
ylabel('y', 'FontSize', 11);
title('(a) Exact solution u_{exact}(x, y)', 'FontSize', 11);
axis equal;
xlim([-1, 1]); ylim([-1, 1]);
caxis([vmin, vmax]);
colorbar;
colormap(gca, 'jet');

% Panel (b): Numerical solution
subplot(2, 2, 2);
contourf(xx, yy, U, 30, 'LineStyle', 'none');
hold on;
contour(xx, yy, U, 10, 'k', 'LineWidth', 0.3);
xlabel('x', 'FontSize', 11);
ylabel('y', 'FontSize', 11);
title(sprintf('(b) Numerical solution u_N(x, y) at N = %d', N), 'FontSize', 11);
axis equal;
xlim([-1, 1]); ylim([-1, 1]);
caxis([vmin, vmax]);
colorbar;
colormap(gca, 'jet');

% Panel (c): Pointwise log10 error
subplot(2, 2, 3);
err_plot = log10(max(err, 1e-16));
contourf(xx, yy, err_plot, 20, 'LineStyle', 'none');
xlabel('x', 'FontSize', 11);
ylabel('y', 'FontSize', 11);
title(sprintf('(c) log_{10}|u_N - u_{exact}| (max = %.2e)', max_error), 'FontSize', 11);
axis equal;
xlim([-1, 1]); ylim([-1, 1]);
cb = colorbar;
ylabel(cb, 'log_{10}|u_N - u_{exact}|', 'FontSize', 9);
colormap(gca, 'parula');

% Panel (d): NEW spectral convergence
subplot(2, 2, 4);
semilogy(N_values, errs_conv + 1e-18, '-o', 'Color', NAVY, ...
         'LineWidth', 1.2, 'MarkerSize', 6, 'MarkerFaceColor', 'w', ...
         'DisplayName', 'max |u_N - u_{exact}|');
hold on;
yline(1e-15, ':', 'Color', [0.5 0.5 0.5], 'LineWidth', 0.6, ...
      'DisplayName', 'machine epsilon');
hold off;
xlabel('N (Chebyshev degree, each direction)');
ylabel('max-norm error');
title('(d) spectral convergence: error vs N', 'FontSize', 11);
grid on; box on;
legend('Location', 'northeast', 'FontSize', 9);

sgtitle('2D Poisson Equation: Spectral Solution', 'FontSize', 13);

%% Save figure
exportgraphics(fig, output_file, 'ContentType', 'vector');
exportgraphics(fig, strrep(output_file, '.pdf', '.png'), 'Resolution', 300);

fprintf('Figure saved to: %s\n', output_file);

close(fig);

%% Local functions

function lap_u = exact_laplacian(xx, yy)
% Analytical Laplacian of the manufactured solution.
% u = g(x)*h(y), g(x) = (1-x^2)*cos(pi*x/2), h(y) = (1-y^2)*cos(pi*y/2).
% Lap(u) = g''(x)*h(y) + g(x)*h''(y).
    p = pi;
    cx = cos(p*xx/2); cy = cos(p*yy/2);
    sx = sin(p*xx/2); sy = sin(p*yy/2);
    gx  = (1 - xx.^2) .* cx;
    hy  = (1 - yy.^2) .* cy;
    gpp = -(2 + p^2/4 * (1 - xx.^2)) .* cx + 2*p*xx .* sx;
    hpp = -(2 + p^2/4 * (1 - yy.^2)) .* cy + 2*p*yy .* sy;
    lap_u = gpp .* hy + gx .* hpp;
end

File: codes/julia/ch10/poisson2d_cheb.jl
# poisson2d_cheb.jl
#
# Chapter 10: Spectral PDE Solvers with Chebyshev and Fourier Grids
#
# 2D Poisson equation solver on Chebyshev tensor grid.
#
# PDE: -Lap(u) = f, -1 < x, y < 1
# BCs: u = 0 on boundary (Dirichlet)
#
# Uses manufactured solution:
#     u(x, y) = (1 - x^2)(1 - y^2) cos(pi*x/2) cos(pi*y/2)
#
# which automatically satisfies homogeneous Dirichlet BCs.
#
# Generates Figure 10.7: Exact, numerical, and error comparison.
#
# Author: Dr. Denys Dutykh
#         Mathematics Department
#         Khalifa University of Science and Technology
#         Abu Dhabi, UAE
#
# Part of the book "Computational Etudes: A Spectral Approach"
# https://github.com/dutykh/computational-etudes

using CairoMakie
using LaTeXStrings
using Colors
using LinearAlgebra
using Printf

# Import cheb_matrix
include(joinpath(@__DIR__, "chebfft.jl"))
include(joinpath(@__DIR__, "..", "ch07", "cheb_matrix.jl"))

# -----------------------------------------------------------------------------
# Publication-quality CairoMakie configuration
# -----------------------------------------------------------------------------
set_theme!(Theme(
    fontsize = 11,
    fonts = (regular = "CMU Serif", bold = "CMU Serif Bold", italic = "CMU Serif Italic"),
    Axis = (xlabelsize = 11, ylabelsize = 11, titlesize = 12,
            xticklabelsize = 10, yticklabelsize = 10,
            spinewidth = 0.8, xtickwidth = 0.8, ytickwidth = 0.8),
    Legend = (labelsize = 10,),
))

# -----------------------------------------------------------------------------
# Configuration
# -----------------------------------------------------------------------------
const NAVY = colorant"#142D6E"

# Output path
SCRIPT_DIR = @__DIR__
OUTPUT_DIR = joinpath(SCRIPT_DIR, "..", "..", "..", "textbook", "figures", "ch10", "julia")
mkpath(OUTPUT_DIR)

# -----------------------------------------------------------------------------
# Manufactured solution
# -----------------------------------------------------------------------------
"""
    exact_solution(xx, yy)

Manufactured exact solution that satisfies u = 0 on the boundary.
"""
exact_solution(xx, yy) = (1 .- xx.^2) .* (1 .- yy.^2) .* cos.(pi .* xx ./ 2) .* cos.(pi .* yy ./ 2)

"""
    exact_laplacian(xx, yy)

Analytical Laplacian of the manufactured solution.

`u = g(x)*h(y)` with `g(x) = (1-x^2)*cos(pi*x/2)`, `h(y) = (1-y^2)*cos(pi*y/2)`.
`Lap(u) = g''(x)*h(y) + g(x)*h''(y)`.
"""
function exact_laplacian(xx, yy)
    p = π
    cx = cos.(p .* xx ./ 2); cy = cos.(p .* yy ./ 2)
    sx = sin.(p .* xx ./ 2); sy = sin.(p .* yy ./ 2)
    gx  = (1 .- xx.^2) .* cx
    hy  = (1 .- yy.^2) .* cy
    gpp = -(2 .+ p^2/4 .* (1 .- xx.^2)) .* cx .+ 2p .* xx .* sx
    hpp = -(2 .+ p^2/4 .* (1 .- yy.^2)) .* cy .+ 2p .* yy .* sy
    return gpp .* hy .+ gx .* hpp
end

# -----------------------------------------------------------------------------
# Poisson solver
# -----------------------------------------------------------------------------
"""
    poisson2d_cheb(; N=32)

Solve 2D Poisson equation with manufactured solution.

The discrete system is:
    A * u_vec = f_vec
where A is the negative Kronecker sum of D2_int with itself.

Returns `(xx, yy, U, U_exact, error)`.
"""
function poisson2d_cheb(; N=32)
    # Build differentiation matrices
    D, x = cheb_matrix(N)
    D2 = D * D
    xx = [x[i] for i in 1:N+1, j in 1:N+1]
    yy = [x[j] for i in 1:N+1, j in 1:N+1]

    # Exact solution
    U_exact = exact_solution(xx, yy)

    # Interior indices
    ii = 2:N
    D2_int = D2[ii, ii]
    I_int = Matrix{Float64}(I, N - 1, N - 1)

    # Kronecker sum Laplacian for interior (represents Delta)
    L = kron(D2_int, I_int) + kron(I_int, D2_int)

    # For -Delta u = f, we have -L * u = f, so use A = -L
    A = -L

    # RHS from analytical Laplacian (not discrete, to test convergence)
    f_full = -exact_laplacian(xx, yy)
    f_int = f_full[ii, ii]

    # Solve A * u = f where A = -Delta
    u_vec = A \ vec(f_int)

    # Reconstruct full solution
    U = zeros(N + 1, N + 1)
    U[ii, ii] = reshape(u_vec, N - 1, N - 1)

    # Compute error
    error = maximum(abs.(U .- U_exact))

    return xx, yy, U, U_exact, error
end

# -----------------------------------------------------------------------------
# Main visualization
# -----------------------------------------------------------------------------
function main()
    # Solve Poisson equation
    N = 32
    xx, yy, U, U_exact, error = poisson2d_cheb(N=N)

    @printf("N = %d, Max error = %.2e\n", N, error)

    # Convergence study upfront so panel (d) can plot it
    println("\nConvergence Study:")
    println("-" ^ 40)
    @printf("%6s  %14s  %10s\n", "N", "Max Error", "Ratio")
    println("-" ^ 40)

    Ns_conv = [8, 12, 16, 20, 24, 28, 32, 40, 48]
    errs_conv = Float64[]
    prev_error = nothing
    for Nv in Ns_conv
        _, _, _, _, err_val = poisson2d_cheb(N=Nv)
        push!(errs_conv, err_val)
        if prev_error !== nothing && err_val > 1e-15
            ratio = prev_error / err_val
        else
            ratio = NaN
        end
        @printf("%6d  %14.6e  %10.2f\n", Nv, err_val, ratio)
        prev_error = err_val
    end

    # Create 2x2 figure
    fig = Figure(size = (1100, 900))

    vmin, vmax = minimum(U_exact), maximum(U_exact)
    n_levels = 30

    # ---- (a) Exact solution -----------------------------------------
    ax1 = Axis(fig[1, 1],
               xlabel = L"x", ylabel = L"y",
               title  = "(a) Exact solution u_exact(x, y)",
               aspect = DataAspect())
    contourf!(ax1, xx[:, 1], yy[1, :], U_exact,
              levels = range(vmin, vmax, length=n_levels), colormap = :RdBu)
    contour!(ax1, xx[:, 1], yy[1, :], U_exact,
             levels = 10, color = (:black, 0.5), linewidth = 0.3)

    # ---- (b) Numerical solution -------------------------------------
    ax2 = Axis(fig[1, 2],
               xlabel = L"x", ylabel = L"y",
               title  = "(b) Numerical solution u_N(x, y) at N = $N",
               aspect = DataAspect())
    contourf!(ax2, xx[:, 1], yy[1, :], U,
              levels = range(vmin, vmax, length=n_levels), colormap = :RdBu)
    contour!(ax2, xx[:, 1], yy[1, :], U,
             levels = 10, color = (:black, 0.5), linewidth = 0.3)

    # ---- (c) Pointwise log10 error ----------------------------------
    ax3 = Axis(fig[2, 1],
               xlabel = L"x", ylabel = L"y",
               title  = "(c) log₁₀ |u_N - u_exact|  (max = $(@sprintf("%.2e", error)))",
               aspect = DataAspect())
    err = abs.(U .- U_exact)
    err_plot = log10.(max.(err, 1e-16))
    contourf!(ax3, xx[:, 1], yy[1, :], err_plot,
              levels = 20, colormap = :viridis)

    # ---- (d) NEW: spectral convergence ------------------------------
    ax4 = Axis(fig[2, 2],
               xlabel = "N (Chebyshev degree, each direction)",
               ylabel = "max-norm error",
               yscale = log10,
               title  = "(d) spectral convergence: error vs N")
    scatterlines!(ax4, Ns_conv, errs_conv .+ 1e-18; color = NAVY,
                  markercolor = :white, strokecolor = NAVY, strokewidth = 1.0,
                  markersize = 6, linewidth = 1.2,
                  label = "max |u_N - u_exact|")
    hlines!(ax4, [1e-15]; color = :gray, linestyle = :dot, linewidth = 0.6,
            label = "machine epsilon")
    axislegend(ax4; position = :rt, labelsize = 9)

    Label(fig[0, :], "2D Poisson Equation: Spectral Solution", fontsize = 14)

    # Save figure
    save(joinpath(OUTPUT_DIR, "poisson2d_solution.pdf"), fig)
    save(joinpath(OUTPUT_DIR, "poisson2d_solution.png"), fig, px_per_unit = 4)
    println("Figure saved to: $(joinpath(OUTPUT_DIR, "poisson2d_solution.pdf"))")
end

main()
